### ISSUE TITLE
vmstorage may have memory leak

### ISSUE THREAD

**Issue by flashmouse** [VictoriaMetrics/VictoriaMetrics#490]
Hi.
we now upgrade vm-cluster to version v1.34.3-cluster because recently we found vmstorage's memory usage increase very fast. even after restart process it will use memoery from 30% to 90% in several days. 
after some research, we believe the root cause is this line in function func (mr *MetricRow) Unmarshal(src []byte) ([]byte, error):
`mr.MetricNameRaw = append(mr.MetricNameRaw[:0], metricNameRaw...)`

MetricRow[] kept in processVMInsertConn function, it will not release until this connection from vminsert closed. we have test to restart vminsert, vmstorage memory usage drop down obviously . i think our client may have some metrics have very long metricname or very long tagvalues text length, it cause MetricRow's byte slice MetricNameRaw use much more memory and won't release,  and if vmstorage run enough time, this array will have more and more such byte slice.

we now simply move `var mrs []storage.MetricRow` to the for loop body, now after 2 month, it seems ok. memory usage now keep about 50% with slight increase.
```
func (s *Server) processVMInsertConn(r io.Reader) error {
	sizeBuf := make([]byte, 8)
	var buf []byte
	//var mrs []storage.MetricRow
	for {
                //move to here
		var mrs []storage.MetricRow
```

i guess other code have such byte slice structure may have this problem too, but in our usage scenario not so serious.

**Comment by valyala** [comment]
@flashmouse , thanks for the detailed analysis of the issue and for the proposed solution! I enhanced the proposed solution in the commit a853869e75de7a5a3e9ed6591211303a7a66eaf0 . The enhanced solution should allocate lower amounts of memory per each accepted packet from `vminsert` on average.

Could you try building cluster components from the commit a853869e75de7a5a3e9ed6591211303a7a66eaf0 and verify whether `vmstorage` stops leaking memory? Note that the upgrade from `v1.34.*` to `v1.35.*` and to the latest commits in the `cluster` branch could result in error logs in `vminsert` and `vmstorage` components during the upgrade because of the modified communication protocol. These errors should stop appearing after all the `vminsert` and `vmstorage` components are upgraded.

**Comment by valyala** [comment]
The bugfix has been included in [v1.35.5](https://github.com/VictoriaMetrics/VictoriaMetrics/releases) release. Closing the issue as fixed.

@flashmouse , feel free re-opening the issue if `v1.35.5-cluster` doesn't fix the memory leak.
